### PR TITLE
Add DBaaS Metrics for connections and inventories

### PR DESCRIPTION
**Pull request by jianrongzhang89**
Sample metrics from testing:

# HELP dbaas_connection_creation_ready_seconds Elapsed time from DBaaS provider connection creation to ready
# TYPE dbaas_connection_creation_ready_seconds gauge
dbaas_connection_creation_ready_seconds{connection="dbaas-cluster1-608df625aa94426b4167e45f",namespace="openshift-dbaas-operator",provider="Red Hat DBaaS / MongoDB Atlas"} 1
# HELP dbaas_connection_status_ready DBaaS provider connection is ready for binding
# TYPE dbaas_connection_status_ready gauge
dbaas_connection_status_ready{connection="dbaas-cluster1-608df625aa94426b4167e45f",namespace="openshift-dbaas-operator",provider="Red Hat DBaaS / MongoDB Atlas"} 1
# HELP dbaas_connection_status_reason Connection status reason
# TYPE dbaas_connection_status_reason gauge
dbaas_connection_status_reason{connection="dbaas-cluster1-608df625aa94426b4167e45f",namespace="openshift-dbaas-operator",provider="Red Hat DBaaS / MongoDB Atlas",reason="Ready"} 1
# HELP dbaas_inventory_creation_ready_seconds Elapsed time from DBaaS provider inventory creation to sync ready
# TYPE dbaas_inventory_creation_ready_seconds gauge
dbaas_inventory_creation_ready_seconds{inventory="my-db-provider-account",namespace="openshift-dbaas-operator",provider="Red Hat DBaaS / MongoDB Atlas"} 53
# HELP dbaas_inventory_status_ready DBaaS provider inventory is sync ready
# TYPE dbaas_inventory_status_ready gauge
dbaas_inventory_status_ready{inventory="my-db-provider-account",namespace="openshift-dbaas-operator",provider="Red Hat DBaaS / MongoDB Atlas"} 1
# HELP dbaas_inventory_status_reason Inventory status reason
# TYPE dbaas_inventory_status_reason gauge
dbaas_inventory_status_reason{inventory="my-db-provider-account",namespace="openshift-dbaas-operator",provider="Red Hat DBaaS / MongoDB Atlas",reason="SyncOK"} 1
# HELP dbaas_platform_status status of an installation of components and provider operators
# TYPE dbaas_platform_status gauge
dbaas_platform_status{platform="crunchy-bridge",status="in progress"} 1
# HELP dbaas_request_duration_seconds Duration of upstream calls to provider operator/service endpoints
# TYPE dbaas_request_duration_seconds histogram
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.005"} 0
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.01"} 0
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.025"} 0
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.05"} 0
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.1"} 0
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.25"} 1
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="0.5"} 1
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="1"} 1
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="2.5"} 1
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="5"} 2
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="10"} 2
dbaas_request_duration_seconds_bucket{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas",le="+Inf"} 2
dbaas_request_duration_seconds_sum{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas"} 3.302491699
dbaas_request_duration_seconds_count{action="list_inventories",instance_name="my-db-provider-account",instance_type="DBaaSInventory",outcome="success",provider_name="Red Hat DBaaS / MongoDB Atlas"} 2

### CODE REVIEW

**Comment by redhatHameed** [comment]
@jianrongzhang89 can rebase and remove the conflicts the PR thanks

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (redhatHameed) [inline .vscode/launch.json:1]: remove?

File: config/default/manager_auth_proxy_patch.yaml [hunk 2]
```diff
@@ -1,26 +0,0 @@
-# This patch inject a sidecar container which is a HTTP proxy for the
```
Comment (redhatHameed) [inline config/default/manager_auth_proxy_patch.yaml:1]: why removed?

File: config/prometheus/monitor.yaml [hunk 3]
```diff
@@ -10,11 +10,7 @@ metadata:
 spec:
   endpoints:
```
Comment (redhatHameed) [inline config/prometheus/monitor.yaml:11]: we should not change the existing monitor.YAML, I have created another file for insecure, maybe in production we need to revert this change

File: config/rbac/auth_proxy_client_clusterrole.yaml [hunk 4]
```diff
@@ -1,9 +0,0 @@
-apiVersion: rbac.authorization.k8s.io/v1
-kind: ClusterRole
```
Comment (redhatHameed) [inline config/rbac/auth_proxy_client_clusterrole.yaml:2]: why removed?

File: controllers/dbaasconnection_controller.go [hunk 5]
```diff
@@ -181,3 +195,67 @@ func (r *DBaaSConnectionReconciler) deploymentMutateFn(connection *v1alpha1.DBaa
 		return nil
 	}
 }
+
+var connectionReasonCache metrics.ProviderReasonsCache = metrics.ProviderReasonsCache{}
```
Comment (redhatHameed) [inline controllers/dbaasconnection_controller.go:199]: is this possible to move this inside metric.go file

File: controllers/dbaasinventory_controller.go [hunk 6]
```diff
@@ -200,3 +217,67 @@ func inventoryRbacObjs(inventory v1alpha1.DBaaSInventory, tenantList v1alpha1.DB
...
+
+func setInventoryStatusReason(inv *v1alpha1.DBaaSInventory, reason, provider string) {
+	inventoryReasonCache.SetProviderReason(provider, reason)
+	resetInventoryStatusReasons(inv, provider)
+	metrics.InventoryStatusReason.With(prometheus.Labels{
+		"provider":  provider,
+		"inventory": inv.Name,
+		"namespace": inv.Namespace,
+		"reason":    reason}).Set(1)
+}
```
Comment (redhatHameed) [inline controllers/dbaasinventory_controller.go:283]: can we move this code inside metrics?go file

File: controllers/metrics/metrics.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+	PlatformStatus,
+	InventoryElapsedTime,
+	InventoryStatusReady,
+	InventoryStatusReason,
+	ConnectionElapsedTime,
+	ConnectionStatusReady,
+	ConnectionStatusReason,
+}
+
+func init() {
```
Comment (redhatHameed) [inline controllers/metrics/metrics.go:111]: can we move into main?go init function

**Review by redhatHameed — COMMENTED** [review]
added some comments, however we want to see this how it looks on dashboard thanks

**Comment by redhatHameed** [comment]
@jianrongzhang89 is that possible to rebase this

**Comment by jianrongzhang89** [comment]
> @jianrongzhang89 can rebase and remove the conflicts the PR thanks

@redhatHameed Sure.
